python/snewpy/rate_calculator.py

```diff
@@ -13,11 +13,281 @@
+            channel name
+        flavor:Flavor or list[Flavor]
+            flavor of the interacting neutrino
+        xsec:callable or FunctionOfEnergy
+            crossection as a function of energy
+        smearing:SmearingMatrix
+            The detection energy smearing matrix. 
+            If `None`(default) no smearing is applied
+        efficiency: float or np.ndarray or callable.
+            The detection efficiency vs. detected energy. 
+            If scalar (float), efficiency is independent of energy. Value of `1.` (default) means 100% efficiency.
+            If it's a vector (np.ndarray), the values correspond to the efficiency in each bin of `smearing.bins_smeared`.
+            If it's a callable it should define efficiency as a function of energy
+        weight:float.
+            Channel weight, to be multiplied to the resulting event rates.
+            Default value is 1.
+        """
+        self.flavor=flavor
+        self.xsec = xsec
+        self.efficiency = efficiency
+        self.smearing = smearing
+        self.weight = weight
+
+    @property 
+    def flavor(self):
+        return self._flavor
+    @flavor.setter
+    def flavor(self, flavors):
+        if isinstance(flavors, Flavor):
+            flavors = [flavors]
+        self._flavor = flavors
+    @property
+    def efficiency(self):
+        return self._efficiency
+        
+    @efficiency.setter
+    def efficiency(self, efficiency):
+        if callable(efficiency) and not isinstance(efficiency,FunctionOfEnergy):
+            efficiency = FunctionOfEnergy(efficiency)
+        self._efficiency = efficiency
+
+    @property
+    def xsec(self):
+        return self._xsec
+    @xsec.setter
+    def xsec(self, xsec):
+        if callable(xsec) and not isinstance(xsec,FunctionOfEnergy):
+                xsec = FunctionOfEnergy(xsec)
+        self._xsec = xsec
+    
+    def __repr__(self):
+        return f'{self.__class__.__name__} (flavor={",".join([f.name for f in self.flavor])}, smearing={self.smearing is not None}, weight={self.weight})'
+    def calc_rate(self, flux:Container, apply_smearing=True, apply_efficiency=True)->Container:
+        """Calculate the event rate in this channel
+        
+        Parameters
+        ----------
+        flux:Container
+            Input neutrino flux
+        apply_smearing:bool
+            If `True`(default) apply the energy smearing, otherwise the result will be just interaction rate vs. neutrino energy
+        apply_efficiency:bool
+            If `True`(default) apply the efficiency, otherwise assume 100% efficiency
+        """
+        rate = self._calc_interaction_rate(flux)
+        if apply_smearing:
+            if self.smearing is not None:
+                rate = self.smearing.apply(rate)
+        if apply_efficiency:
+            if isinstance(self.efficiency, FunctionOfEnergy):
+                rate = self.efficiency * rate
+            else:
+                rate = rate*self.efficiency
```

Comment:
Here both `if` and `else` cases give same rate, is this condition needed?

Reply:
You're right! That was initially done because `rate` contains a 3D array (Flavor,Time,Energy) and multiplication order matters, if efficiency is a vector (`rate*efficiency` will work correctly, on the Energy axis), and for the `FunctionOfEnergy` it's the class method `__mul__` which does all the work. 
But it'll be better if I get rid of this condition here, and for all cases it will be `rate = rate*self.efficiency`.

Below is just explanations of what I did in the commits, to achieve this:

1. Created `FunctionOfEnergy.__rmul__` method, so that it can be multiplied from the right.
2. Restricted the `flux.Container.__mul__` mehtod: if it's multiplied not by scalar of NumPy array, this returns `NotImplemented`. That means if I multiply by `FunctionOfEnergy` if will not use `Container.__mul__`, but instead `FunctionOfEnergy.__rmul__`

Reply:
okey, it is more clear now, thanks for explaining